### DISCUSSION TITLE
pb.collection is not a function

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#1070] [Q&A]
Ello Pocketbase users / maintainers! :D first of all thank you for your hard work on this project.

Second of all... :sweat_smile: 
I've migrated my pocketbase instance from 0.7.x to 0.8.0 and the JS SDK is giving me problems with svelte. All of my components died x.x and I figured out I could just make a tiny wrapper to avoid repeating a lot of code.

(I've tested this approach on both a standalone component and an external file as a "helper" function):

```typescript
import Pocketbase from 'pocketbase';
import { fapi_url, api_user, api_user_pass } from '../../config';


/** Reusable function to fetch items from PocketBase */
export async function fetch_items(table: string, qty: number) {
    const pb: Pocketbase = new Pocketbase(fapi_url);

    pb.collection('users').authWithPassword(api_user, api_user_pass);
    let result = await pb.collection(table).getFullList(qty, {
        sort: '-created',
    });
    pb.authStore.clear();

    return result;
}
```

However, both the IDE and the browser report the following error:

`Uncaught (in promise) TypeError: pb.collection is not a function`

Am I doing something wrong? where did I mess up? :thinking: my `package.json` file has this entry for pocketbase: `"pocketbase": "^0.8.0",` is my syntax wrong? am I missing something? :thinking:  I've tried copypasting the hints from my pocketbase instance but it throws the same error over and over :(

**Comment by ganigeorgiev** [comment]
Are you sure that you are using PocketBase JS SDK v0.8? Try running again `npm update`.

**Reply by ghost** [reply]
I restarted the server and some errors are not logged now, however typescript complains a lot at IDE level. I think it doesn't die at runtime because of ESBuild and it's type-skip:

`Property 'getFileUrl' does not exist on type 'Client'.ts(2339)` is the most reported error in here :thinking: 

This also happens with `collection`:

`Property 'collection' does not exist on type 'Client'. Did you mean 'collections'?ts(2551)`

**Reply by ganigeorgiev** [reply]
I don't know what could be the reason for the above errors (it may be possible that your text editor has cached the old TS definitions).

**Reply by ganigeorgiev** [reply]
Both `collection` and `getFileUrl` could be found in the generated TS definitions file so most likely is some strange cache issue:
- https://github.com/pocketbase/js-sdk/blob/master/dist/pocketbase.es.d.ts#L755
- https://github.com/pocketbase/js-sdk/blob/master/dist/pocketbase.es.d.ts#L777

As a last resort try deleting `node_modules`, run again `npm update` and restart your text editor.

**Reply by ghost** [reply]
Okay that seemed to fix it. Deleting node modules, re-installing everyting and reloading the editor was all that was needed to fix this. I thought vite would reload definitions / newpackages by itself, but it seems that it isn't the case. :bow:  Thank you so much!

BTW: To anyone reading, I opted for the component repeated-code method since my function returned a promise and that isn't iterable by `{#each}`.

**Comment by ghost** [comment]
Reopening as this happens on Angular 14 as well, did all of the above but I get a TypeScript errors telling me that:

`Property 'collection' does not exist on type 'Client'. Did you mean 'collections'?`

**Reply by ghost** [reply]
I've reinstalled all of the node_modules, refreshed the editor and even restarted my computer to no avail :(

**Reply by ganigeorgiev** [reply]
If you are sure that you are using PocketBase JS SDK v0.8.0, then maybe there is a bug with the editor/lsp plugin.

Another possibility for the error could be also if you have multiple projects in a monorepo and some of them use an older PocketBase version and the editor is wrongly picking its definition instead of each individual project deps.

But in any case I don't think this is PocketBase related and you can try lookup for similar issues in the editor/typescrtipt/lsp-plugin repos you are using. For example, in the TypeScript repo there is an open issue that describes stale definition error similar to yours - https://github.com/microsoft/TypeScript/issues/48981.

**Reply by ghost** [reply]
I suspect it might be a way I'm typing my variables or if TypeScript and it's "dynamics" are messing things up.

If I type my new object as `any` instead of `PocketBase` or letting "type inference" do it's work:

```typescript
const pb = new Pocketbase(environment.apiHost);  
// After
const pb: any = new Pocketbase(environment.apiHost);  
```

The error goes away on the editor, however the chrome console does report it:

![image](https://user-images.githubusercontent.com/24773698/203245290-fc3f909e-2cb6-42ca-97cf-250e99331c94.png)

For interests sake here are the two versions of the code that don't work:

EDIT: Made a typo of "Pocketbase" instead of "PocketBase" (when copying here)

```typescript
  getAllObjects(table: string) {
    // Using the IDE suggestion "PocketBase" or leaving the type blank
    const pb = new PocketBase(environment.apiHost);  
    pb.collection('users').authWithPassword(environment.user, environment.password);
    const res = pb.collection(table).getFullList(200, {
      sort: '-created',
    });
    pb.auth_data.clear();

    console.log(res);
    return res;
  }
```

and

```typescript
  getAllObjects(table: string) {
    // Ignoring TS and just typing any as the type
    const pb: any = new PocketBase(environment.apiHost);  
    pb.collection('users').authWithPassword(environment.user, environment.password);
    const res = pb.collection(table).getFullList(200, {
      sort: '-created',
    });
    pb.auth_data.clear();

    console.log(res);
    return res;
  }
```

**Reply by ganigeorgiev** [reply]
If the error is in the browser, then I suspect that you are not using PocketBase JS SDK v0.8.0. Please check again your `package.json`.

**Reply by ghost** [reply]
My package.json shows `"pocketbase": "^0.8.0",` :thinking: I tried installing the sdk on a blank project with `npm install -D pocketbase` and it still installs `0.7.x`

At this point im starting to question my node installation :upside_down_face:
